### PR TITLE
fix(compactor): retry OOM error at reduced concurrency

### PR DESCRIPTION
**Pull request by Joe-Blount**
This is related to some of the partitions left uncompacted in:
https://github.com/influxdata/EAR/issues/4505
https://github.com/influxdata/EAR/issues/4496

When the compactor gets a ResourcesExhausted / OutOfMemory error from DF, we can retry at reduced concurrency.  

When DF needs more than the configured limit, it returns an error to the compactor.  With this PR, the compactor will retry the compaction requesting a higher number of permits from the DF job semaphore (so this one job takes a higher share of the total concurrency).  The effect is that the job is retried while allowing fewer jobs to run concurrently.  The retries will repeat until it passes, or it is forced to run alone and still fails with resource exhausted.

When we get a separate scheduler we'll have more options for responding to these errors, but for now this is an easy change that will improve most situations.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/src/driver.rs [hunk 1]
```diff
@@ -556,76 +556,98 @@ async fn execute_plan(
...
+            drop(df_span);
 
-        // inputs can be removed from the scratchpad as soon as we're done with compaction.
+            info!(
+                partition_id = partition_info.partition_id.get(),
+                plan_id, "job semaphore released",
+            );
+
+            if let Err(e) = &res {
+                if e.classify() == ErrorKind::OutOfMemory {
```
Comment (Joe-Blount) [inline compactor/src/driver.rs:630]: I tried to put these in a single `if`, but the compiler says `if let` followed by more conditions is unstable :frowning_face:

Comment (wiedld) [reply compactor/src/driver.rs:630]: suggestion: could you use a match?

File: compactor/src/driver.rs [hunk 2]
```diff
@@ -556,76 +556,98 @@ async fn execute_plan(
...
+                break;
+            }
+        }
+
+        // inputs can be removed from the scratchpad as soon as we're done with compaction
         scratchpad_ctx
             .clean_from_scratchpad(&plan_ir.input_paths())
             .await;
 
-        info!(
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/src/driver.rs:624]: I thought about that, but the `job semaphore released` corresponds to `requesting job semaphore` and `job semaphore acquired`.  Keeping those as a matched set seems cleaner (and easier to script checks for).  

The added log is getting a bit verbose, but its rare, so extra logging in the rare circumstances isn't concerning.

File: compactor/src/driver.rs (hunk 1 above)
Comment (Joe-Blount) [reply compactor/src/driver.rs:630]: yeah, that looks a little better..

**Pull request merged**